CONTRACTS: Front-end extensions for assigns clauses

remi-delmas-3000 opened the pull request:
The first commit tags some JBMC tests as THOROUGH because they seem to exceed resources available for the ubuntu 18.04 target when building this PR

Add new functions to specify assignable targets in assigns clauses:
- add polymorhpic builtin __CPROVER_typed_target
- add builtin __CPROVER_assignable
- add builtin __CPROVER_object_whole
- add builtin __CPROVER_object_from
- add builtin __CPROVER_object_upto
- allow function call expressions as assigns
  clause targets as long as they are one of
  the new built-ins

Using __CPROVER_POINTER_OBJECT in assigns clauses is still allowed for now (needed for loop contracts),
will be deprecated in #7127.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [x] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [n/a] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

feliperodri commented on doc/cprover-manual/contracts-assigns.md:
````diff
@@ -17,28 +17,46 @@ value(s) therein are not modified.
+ of `sizeof(expr_t)` bytes starting at `&expr` is assignable:
 ```c
-__CPROVER_size_t __CPROVER_object_from(void *ptr); 
+__CPROVER_assignable_t __CPROVER_typed_target(expr_t expr);
````
In this section, we do not explain the meaning of `*ptr` in assigns clause, which seems to be equivalent to `__CPROVER_typed_target`. For instance,
```c
void foo(int a, struct bar *b, char *c)
__CPROVER_assigns(__CPROVER_assignable_t(b))
__CPROVER_assigns(*c)
;
```
What is the difference between these two assigns clauses? Here is another good place to remember the user  everything in C is pass-by-value and we don't need to include these regions in the assigns clause (e.g., `a`).

remi-delmas-3000 replied:
They are not equivalent:

`__CPROVER_assigns(__CPROVER_typed_target(b))` says the range of bytes starting at `&b` up to `sizeof(b)` is assignable, which is basically the pointer `b` itself.

`__CPROVER_assigns(*b)` says the range of bytes starting at `&(*b)` up to `sizeof(*b)` is assignable, so that is the struct.

remi-delmas-3000 replied:
The documentation gives the start address and the size of the range in bytes for each builtin operator, what additional clarification do you think we could add ?

feliperodri commented on doc/cprover-manual/contracts-assigns.md:
```diff
@@ -17,28 +17,46 @@ value(s) therein are not modified.
+call-by-contract replacement, `__CPROVER_havoc_slice(ptr, size)` is used to
+havoc these byte ranges, and `__CPROVER_havoc_slice` does not support
+havocing pointers. `__CPROVER_typed_target` must be used to specify targets
+that are pointers.
```
We should also improve our examples below to make use of these expressions or at least provide a few concrete examples of how to use each one of them.

remi-delmas-3000 replied:
I added an example in the doc. The assigns clause documentation (and the frees clauses as well as is_fresh) is going to be rewritten for the dynamic frames PR anyway so I'd like to invest on the new doc instead of patching the old one.

feliperodri commented on doc/cprover-manual/contracts-assigns.md:
```diff
@@ -17,28 +17,46 @@ value(s) therein are not modified.
 
-The following functions can be used in assigns clause to specify ranges of 
-assignable addresses.
+The following functions can be used in assigns clauses to specify ranges of assignable bytes.
```
Is there any difference between assigns clauses in function and loop contracts at this point? If no, we don't need two subsections as it is today.

feliperodri commented on doc/cprover-manual/contracts-frees.md:
```diff
@@ -0,0 +1,80 @@
+[CPROVER Manual TOC](../../)
+
+# Frees Clause
```
I'd place everything related to frees clauses on a separate commit.

feliperodri replied:
Better yet, why don't just move everything related to free to the next PR?

feliperodri commented on doc/cprover-manual/contracts-frees.md:
````diff
@@ -0,0 +1,80 @@
+__CPROVER_frees(*pointer-typed-expression*, ...)
+```
+
+A _frees_ clause allows the user to specify a set of pointers that may be freed
````
I'd place this phrase before the subsection "Syntax", cause it's a clear succinct definition for *frees* clauses.

feliperodri commented on doc/cprover-manual/contracts-frees.md:
````diff
@@ -0,0 +1,80 @@
+condition: list-of-pointer-typed-expressions;
+```
+
+### Examples
````
Wouldn't full examples come after "Semantics"?

feliperodri commented on doc/cprover-manual/contracts-frees.md:
```diff
@@ -0,0 +1,80 @@
+int foo(char *arr1, char *arr2, size_t size)
+__CPROVER_frees(
+    // arr1 freeable only if the condition holds
+    size > 0 && arr1: arr1;
```
We don't need to specify the exactly same condition in the *frees* clause to avoid error, right? For instance, if `arg2` is NULL or invalid, would it lead to an error in our *frees* clause?

feliperodri commented on doc/cprover-manual/contracts-frees.md:
```diff
@@ -0,0 +1,80 @@
+
+### Semantics
+
+The set of pointers specified by the frees clause of the contract is interpreted
```
```suggestion
The set of pointers specified by the _frees_ clause of the contract is interpreted
```

feliperodri commented on doc/cprover-manual/contracts-frees.md:
```diff
@@ -0,0 +1,80 @@
+
+#### For contract checking
+When checking a contract against a function or a loop, each pointer that the
+function or loop body attempts to free gets checked for membership in the set of
```
This is the first time I'm seeing we mention "membership". It might be more coherent with the rest of the manual if we say "CBMC checks whether each pointer in the _frees_ clauses belong to the write set of the function", or something like that.

feliperodri commented on doc/cprover-manual/contracts-frees.md:
```diff
@@ -0,0 +1,80 @@
+
+#### For replacement of function calls or loops by contracts
+When replacing a function call or a loop by a contract, each pointer of the
+_frees_ clause is non-deterministically freed after the function call
```
I guess we will have other predicates to specify in the ensures clauses when an object is actually freed, right?

feliperodri commented on src/ansi-c/ansi_c_convert_type.cpp:
```diff
@@ -341,6 +349,9 @@ void ansi_c_convert_typet::write(typet &type)
 
+    if(!frees.empty())
+      to_code_with_contract_type(type).frees() = std::move(frees);
+
```
Documentation should state the semantics of empty *frees* clauses.

feliperodri commented on src/ansi-c/c_typecheck_code.cpp:
```diff
@@ -1078,8 +1119,8 @@ void c_typecheck_baset::typecheck_spec_assigns(exprt::operandst &targets)
       error().source_location = target.source_location();
-      error() << "expected ID_conditional_target_group expression in assigns "
-                 "clause, found "
+      error() << "expected ID_conditional_target_group expression in " +
```
This is for debug purposes of to inform the user an error?

feliperodri commented on src/goto-instrument/contracts/instrument_spec_assigns.cpp:
```diff
@@ -519,6 +519,44 @@ car_exprt instrument_spec_assignst::create_car_expr(
+        return {
+          condition,
+          target,
+          minus_exprt(
+            typecast_exprt::conditional_cast(ptr, pointer_type(char_type())),
+            pointer_offset(ptr)),
+          typecast_exprt::conditional_cast(object_size(ptr), size_type()),
```
Nitpick: a comment above this expression would be helpful. Same below.

nwetzler replied:
+1

feliperodri commented on regression/contracts/frees-clause-for-loop-fail/test.desc:
```diff
@@ -0,0 +1,10 @@
+^SIGNAL=0$
+--
+--
+This test is expected trigger a typchecking error on a frees clause target.
```
```suggestion
This test is expected trigger a type-checking error on a frees clause target.
```

feliperodri commented on regression/contracts/frees-clause-for-loop-pass/test.desc:
```diff
@@ -0,0 +1,9 @@
+^SIGNAL=0$
+--
+--
+This test checks that frees clauses are parsed and typechecked for for loops.
```
```suggestion
This test checks that frees clauses are parsed and typechecked for loops.
```

feliperodri commented on regression/contracts/frees-clause-function-fail/main.c:
```diff
@@ -0,0 +1,19 @@
+#include <stdlib.h>
+
+int foo(char *arr, int size)
+  // clang-format off
```
format

feliperodri commented on regression/contracts/frees-clause-function-fail/test.desc:
```diff
@@ -0,0 +1,10 @@
+^SIGNAL=0$
+--
+--
+This test is expected trigger a typchecking error on a frees clause target.
```
Fix all "typchecking" typos with "type-checking"

feliperodri commented on regression/contracts/frees-clause-function-pass/main.c:
```diff
@@ -0,0 +1,20 @@
+#include <stdlib.h>
+
+int foo(char *arr1, char *arr2, size_t size)
+  // clang-format off
```
Fix all format errors with extra space before `// clang-format off`.

feliperodri commented on regression/contracts/frees-clause-while-loop-fail/frees-clause-while-loop-pass/main.c:
```diff
@@ -0,0 +1,20 @@
+#include <stdlib.h>
+int main()
+{
+  size_t size = 10;
+  char *arr = malloc(size);
+  size_t i = 0;
+  while(i <= size)
+    // clang-format off
+  __CPROVER_assigns(i, arr[2], __CPROVER_POINTER_OBJECT(arr))
+  __CPROVER_frees(arr)
+    // clang-format on
+    {
+      if(i == 2)
+        arr[i] = 0;
+      if(i == size)
+        free(arr);
+      i++;
+    }
+  return 0;
+}
```
You should delete this file.

feliperodri commented on regression/contracts/frees-clause-while-loop-fail/frees-clause-while-loop-pass/test.desc:
```diff
@@ -0,0 +1,9 @@
+CORE
+main.c
+--apply-loop-contracts
+^VERIFICATION SUCCESSFUL$
+^EXIT=0$
+^SIGNAL=0$
+--
+--
+This test checks that frees clauses are parsed and typechecked for while loops.
```
You should delete this file.

feliperodri commented on regression/contracts/reject_history_expr_in_assigns_clause/main.c:
```diff
@@ -0,0 +1,13 @@
+#include <assert.h>
```
There are no asserts. Remove it.

feliperodri commented on regression/contracts/reject_history_expr_in_frees_clause/main.c:
```diff
@@ -0,0 +1,13 @@
+#include <assert.h>
```
This should be `stdlib.h`

feliperodri commented on regression/contracts/reject_history_expr_in_assigns_clause/test.desc:
```diff
@@ -0,0 +1,11 @@
+CORE
+main.c
+--enforce-contract foo
+^main.c:.*error: assigns clause target must be an lvalue or a call to __CPROVER_POINTER_OBJECT or to a function returning __CPROVER_assignable_t$
```
Why this is a different error message? Shouldn't it be the following?
```
^main.c.*error: __CPROVER_old is not allowed in frees clauses.$
```

feliperodri commented on regression/contracts/reject_history_expr_in_preconditions/main.c:
```diff
@@ -0,0 +1,13 @@
+#include <assert.h>
```
No assertions. Remove it.

feliperodri commented on regression/contracts/reject_return_value_in_assigns_clause/main.c:
```diff
@@ -0,0 +1,12 @@
+#include <assert.h>
```
No assertions. Remove it.

feliperodri commented on regression/contracts/reject_return_value_in_frees_clause/main.c:
```diff
@@ -0,0 +1,12 @@
+#include <assert.h>
```
This should be `stdlib.h`.

feliperodri commented on regression/contracts/reject_return_value_in_preconditions/main.c:
```diff
@@ -0,0 +1,12 @@
+#include <assert.h>
```
No assertions. Remove it.

feliperodri commented on regression/contracts/typed_target_fail_wrong_nof_operand/main.c:
```diff
@@ -0,0 +1,12 @@
+#include <assert.h>
```
No assertions. Remove it.

feliperodri commented on regression/contracts/typed_target_fail_wrong_nof_operand/test.desc:
```diff
@@ -0,0 +1,11 @@
+^SIGNAL=0$
+--
+--
+This test checks that __CPROVER_old occurences in assigns clauses
```
Not true, update the description.

feliperodri replied:
+1

feliperodri commented on regression/contracts/typed_target_pointer/main.c:
```diff
@@ -0,0 +1,15 @@
+#include <assert.h>
```
No assertions. Remove it.

feliperodri reviewed (changes requested):
I'd suggest that future PRs could be break down into smaller commits and maybe you wanna move everything related to *frees* clauses to the next PR (which would remove the dependency from the other PR and make both of them much cleaner). Apart from that, only small comments. Great improvement!!!

Commit:
CONTRACTS: Front-end extensions for assigns clauses.

Add new functions to specify assignable
targets in assigns clauses:
- add polymorhpic builtin __CPROVER_typed_target
- add builtin __CPROVER_assignable
- add builtin __CPROVER_object_whole
- add builtin __CPROVER_object_from
- add builtin __CPROVER_object_upto
- allow function call expressions as assigns
  clause targets as long as they are one of
  the new built-ins

Using __CPROVER_POINTER_OBJECT is still allowed (for loop contracts),
will be deprecated in an ulterior PR.

feliperodri commented on src/ansi-c/c_typecheck_code.cpp:
```diff
@@ -847,7 +847,26 @@ void c_typecheck_baset::typecheck_dowhile(code_dowhilet &code)
+    error().source_location = expr.source_location();
+    error() << "side-effects not allowed in assigns clause targets"
+            << messaget::eom;
+    throw 0;
```
Future: https://github.com/diffblue/cbmc/issues/6468

remi-delmas-3000 replied:
now throwing `invalid_source_file_exceptiont`

feliperodri commented on src/ansi-c/c_typecheck_code.cpp:
```diff
@@ -903,39 +922,91 @@ void c_typecheck_baset::typecheck_spec_assigns_condition(exprt &condition)
+    {
+      must_throw = true;
+      error().source_location = target.source_location();
+      error() << "expected ID_conditional_target_group expression in " +
```
Do we want to inform users about `ID_conditional_target_group`? Perhaps we could have a more informative message.

remi-delmas-3000 replied:
The syntax section of the manual defines a conditional target group as :

```
conditional-target-group ::= (condition ':')? target (',' target)*
``` 

How about using `conditional target group` instead of `ID_conditional_target_group` to make it more human-readable ?

feliperodri replied:
I like it!

feliperodri commented on src/ansi-c/c_typecheck_expr.cpp:
```diff
@@ -1943,8 +1943,59 @@ void c_typecheck_baset::typecheck_side_effect_function_call(
     if(symbol_table.symbols.find(identifier)==symbol_table.symbols.end())
     {
       // This is an undeclared function.
+
```
Does it make sense to keep the comment above?

remi-delmas-3000 replied:
seems like this comment has always been here, and has follow-up comments downstream

feliperodri commented on doc/cprover-manual/contracts-assigns.md:
```diff
@@ -1,50 +1,166 @@
+
+### Lvalue targets
+
+An l-value target `expr` with a complete type `expr_t` specifies that the range
```
Nitpick: either all Lvalue or l-value.

feliperodri replied:
+1

remi-delmas-3000 replied:
fixed

feliperodri commented on doc/cprover-manual/contracts-assigns.md:
```diff
@@ -1,50 +1,166 @@
+    buf->current = buf->first;
+  else
+    buf->current += 1;
+}
```
Perfect! 👏

feliperodri commented on doc/cprover-manual/contracts-assigns.md:
````diff
@@ -1,50 +1,166 @@
+`__CPROVER_object_whole(ptr)` specifies that all bytes of the object `o`
+are assignable:
+```c
+__CPROVER_assignable_t __CPROVER_object_whole(void *ptr);
````
We should add an example for `__CPROVER_object_whole`.

remi-delmas-3000 replied:
I added an single example to support all three slice primitives

feliperodri commented on doc/cprover-manual/contracts-assigns.md:
```diff
@@ -1,50 +1,166 @@
+### Inductive data structures
+Inductive data structures are not supported yet in assigns clauses.
 
+## Examples
```
Are these examples still valid? I think we code remove this "Examples Section" completely.

feliperodri replied:
+1

remi-delmas-3000 replied:
Removed the section

feliperodri commented on doc/cprover-manual/contracts-assigns.md:
```diff
@@ -1,50 +1,166 @@
+The target expression must be free of function calls and side-effects.
+The condition expression may contain calls to side-effect-free functions.
+
+### Lvalue targets
```
We need to show what would be equivalent now to say `__CPROVER_assign(*ptr)`.

feliperodri replied:
+1

remi-delmas-3000 replied:
`__CPROVER_assigns(*ptr)` can still be used.

feliperodri replied:
But that is the only one we don't explain the semantics in the text.

remi-delmas-3000 replied:
`*ptr` is an lvalue so the definition is given in the section:

> An l-value target `expr` with a complete type `expr_t` specifies that the range of `sizeof(expr_t)` bytes starting at `&expr` is assignable.

remi-delmas-3000 replied:
Added examples of lvalue targets vs non-lvalue targets.

feliperodri reviewed (approved):
A few more comments but it should be good to go.

tautschnig commented:
Just noting that I posted a bunch of comments in #7091 that apply to the second commit of this PR. Apologies for not noticing that I should have reviewed this one first. Will re-review this one once the updates have been made.

remi-delmas-3000 commented:
@tautschnig the `typedef void __CPROVER_assignable_t` has been removed as suggested.

feliperodri commented on regression/contracts/assigns_enforce_arrays_02/test.desc:
```diff
@@ -4,14 +4,14 @@ main.c
 ^\[f2.assigns.\d+\] line 14 Check that a\[.*0\] is assignable: SUCCESS$
 ^\[f2.assigns.\d+\] line 15 Check that a\[.*5\] is assignable: SUCCESS$
-^\[f2.assigns.\d+\] line 16 Check that POINTER_OBJECT\(\(void \*\)a\) is assignable: SUCCESS$
+^\[f2.assigns.\d+\] line 16 Check that POINTER_OBJECT\(\(.* \*\)a\) is assignable: SUCCESS$
```
Why `POINTER_OBJECT` and not `__CPROVER_object_whole`?

remi-delmas-3000 replied:
The new slice operators are not yet supported for loop contracts so POINTER_OBJECT is still supported alongside the new slice operators.

POINTER_OBJECT is still also still used internally to track objects created by `malloc` and to check objects destroyed by `free`.

The dynamic frames PR will completely drop POINTER_OBJECT in assigns clauses, in the front-end and externally, for function contracts and then for loops.

feliperodri reviewed (approved):
Only minor comments.

feliperodri reviewed (approved):
LGTM! @tautschnig I think this PR is ready for your final review.

tautschnig commented on doc/cprover-manual/contracts-assigns.md:
```diff
@@ -1,75 +1,368 @@
+# Assigns Clauses
 
-## In Function Contracts
+An _assigns clause_ lets the user to specify which memory locations may be
```
s/to //

remi-delmas-3000 replied:
fixed

tautschnig commented on doc/cprover-manual/contracts-assigns.md:
```diff
@@ -1,75 +1,368 @@
 
-## In Function Contracts
+An _assigns clause_ lets the user to specify which memory locations may be
+assigned to by a function or a loop.
```
Perhaps a sentence or two can be added _why_ assigns clauses are useful/required? The above says what they describe, but (naively speaking) life would be easier if one didn't have to write any such clauses.

remi-delmas-3000 replied:
One of our goals is to make contracts standalone and independent of a particular function, in which case there is no programs to infer a clause from.

One way to figure out the side effects of a particular function however could be to check it under an empty assigns clause and to turn all reported violations into an assigns clause.

For loops it is different because they necessarily live in a function and just happen to assign whatever they assign,.

remi-delmas-3000 replied:
fixed

tautschnig commented on doc/cprover-manual/contracts-assigns.md:
```diff
@@ -1,75 +1,368 @@
+For a function contract, if no assigns clause is provided, the default is the
+empty set.
+
+For a loop contract, if no assigns clause is provided, CBMC attempts to infer
```
This blank line is distracting: I was about to comment "and what about loops" to then find the answer in the next paragraph.

tautschnig commented on doc/cprover-manual/contracts-assigns.md:
```diff
@@ -1,75 +1,368 @@
+empty set.
+
+For a loop contract, if no assigns clause is provided, CBMC attempts to infer
+the set of locations assigned by the loop from the loop body,
+and then checks that the inferred set is correct (as if specified by the user).
```
So why couldn't we do this for functions as well? Conversely, is it a desirable feature to have this difference in semantics of an unspecified assigns clause? Should we perhaps have a keyword `auto` that's permitted in loop assigns clauses only?

remi-delmas-3000 replied:
Thinking about it, I think we can do it today for loops because the assigns clause inference, the loop abstraction and the verification of the inferred clause all happen at the same time in the same analysis.

For function contracts we operate either in contract checking or contract replacement mode and we cannot infer check and replace the assigns clause in a single run. 

In a later version we could introduce something like '__CPROVER_assigns(auto)` which would mean "when this contract is used to replace some function call, infer the  set  of memory locations from the code of the function, check that the result is correct in a separate analysis, and use that inferred clause to havoc state at the replacement site" 

One way to figure out the side effects of a particular function could be to check it under an empty assigns clause and to turn all reported violations into an assigns clause.

tautschnig commented on doc/cprover-manual/contracts-assigns.md:
```diff
@@ -1,75 +1,368 @@
+and then checks that the inferred set is correct (as if specified by the user).
+
+For both functions and loop contracts, if more than one assigns clause is
+provided, they are merged into a single clause.
```
Can we talk about a "union" instead of the somewhat underspecified "merged?"

remi-delmas-3000 replied:
I'd say "concatenation" to be really precise

remi-delmas-3000 replied:
fixed

tautschnig commented on doc/cprover-manual/contracts-assigns.md:
```diff
@@ -1,75 +1,368 @@
+For both functions and loop contracts, if more than one assigns clause is
+provided, they are merged into a single clause.
+
+We use the _assigns_ interpretation for these memory locations, which means that
+memory locations that are part of the _frame_ of the function call
+(or the loop) and that are not listed in the assigns clause
+(or the _inferred_ assigns clause for a loop contract) must not be assigned to
+by the function (or the loop), even if they end up holding the same value as they
+held before the function call (or before entering the loop).
+
+Memory locations that are not part of the _frame_, i.e. that are locally stack-
+or heap-allocated during function execution or loop execution can always be
+assigned to by the function or the loop.
```
I think we need to disentangle the introduction of "frame" and the difference between "modifies" and "assigns."

remi-delmas-3000 replied:
removed the references to frames

tautschnig commented on doc/cprover-manual/contracts-assigns.md:
````diff
@@ -1,75 +1,368 @@
+
+```
+targets ::= conditional-target-group (';' conditional-target-group)* ';'?
+conditional-target-group ::= (condition ':')? target (',' target)*
````
Indent?

remi-delmas-3000 replied:
fixed

tautschnig commented on doc/cprover-manual/contracts-assigns.md:
```diff
@@ -1,75 +1,368 @@
+
+### Lvalue targets
+
+Roughly speaking, _Lvalues_ are expressions that are associated with a memory location,
```
Lower-case "lvalues?"

remi-delmas-3000 replied:
fixed

tautschnig commented on doc/cprover-manual/contracts-assigns.md:
```diff
@@ -1,75 +1,368 @@
+
+Examples of lvalues are: `x` if `x` is either a global or local variable identifier,
+`*ptr` if `ptr` is a pointer expression, `ptr[i]` or `ptr + i` if `ptr` is
+pointer variable or an array and `i` is an integer exression, etc.
```
s/exression/expression/

remi-delmas-3000 replied:
fixed

tautschnig commented on doc/cprover-manual/contracts-assigns.md:
````diff
@@ -1,75 +1,368 @@
+meaning: for an lvalue `expr` with a complete type `expr_t`,
+`__CPROVER_typed_target(expr)` specifies that the range of `sizeof(expr_t)`
+bytes starting at `&expr` is assignable:
 
-Given a pointer `ptr` pointing into some object `o`, `__CPROVER_object_from(ptr)` 
-specifies that all bytes starting from the given pointer and until the end of 
-the object are assignable:
 ```c
-__CPROVER_size_t __CPROVER_object_from(void *ptr); 
+void __CPROVER_typed_target(expr_t expr);
 ```
````
This explains when one can use that; when does one _have to_ use it? (In other words: why is this syntax even necessary?)

remi-delmas-3000 replied:
This PR was pulled (somehow artificially) from the larger #6887 that introduces using regular C functions as parameterized assigns clause specifications. So this particular one is meant to be used to specify an lvalue target by calling it from a regular C function, but it can also be used in the assigns clause directly.

 If you don't mind I'd like to keep this here and not have to move it back to #6887, everything will make more sense when finally merged

tautschnig replied:
Fine with me!

tautschnig commented on doc/cprover-manual/contracts-assigns.md:
```diff
@@ -1,75 +1,368 @@
-Given a pointer `ptr` pointing into some object `o`, `__CPROVER_object_from(ptr, size)` 
-specifies that `size` bytes starting from the given pointer and until the end of the object are assignable.
-The `size` value must such that `size <= __CPROVER_object_size(ptr) - __CPROVER_pointer_offset(ptr)` holds:
+In order to specify that a memory location which contents is interpreted as a
+pointer by the program is assignable, one must use the notation
```
s/which contents/the contents of which/

remi-delmas-3000 replied:
fixed

tautschnig commented on doc/cprover-manual/contracts-assigns.md:
```diff
@@ -1,75 +1,368 @@
+`__CPROVER_assigns(__CPROVER_typed_target(ptr))`, as opposed to the slice
+operators `__CPROVER_object_whole`, `__CPROVER_object_from`,
+or `__CPROVER_object_upto`.
+This ensures that during call-by-contract replacement the memory location get
```
s/get/gets/

remi-delmas-3000 replied:
fixed

tautschnig commented on doc/cprover-manual/contracts-assigns.md:
```diff
@@ -1,75 +1,368 @@
-structs, and arrays.  Recursive pointer structures are left to future work, as
-their support would require changes to CBMC's memory model.
+
+The value of `size` must such that the range does not exceed
+the object's boundary, i.e.
+`size <= __CPROVER_OBJECT_SIZE(ptr) - __CPROVER_POINTER_OFFSET(ptr)` must hold.
```
Just curious: what happens when the user violates this rule?

remi-delmas-3000 replied:
an assertion violation happens if that is not the case

remi-delmas-3000 replied:
added precision

tautschnig commented on doc/cprover-manual/contracts-assigns.md:
```diff
@@ -1,75 +1,368 @@
+and of size `__CPROVER_OBJECT_SIZE(ptr) - __CPROVER_POINTER_OFFSET(ptr)` is assignable.
+
+Revisiting our previous example, changing the target to
+`__CPROVER_object_from(vec->data)` still rejects the assignment to vec->size,
```
"vec->size" should be formatted as code.

remi-delmas-3000 replied:
fixed

tautschnig commented on src/ansi-c/ansi_c_internal_additions.cpp:
```diff
@@ -216,6 +216,19 @@ void ansi_c_internal_additions(std::string &code)
+    // frame specifications for contracts
+    // Declares a range of bytes as assignable (internal representation)
+    "void " CPROVER_PREFIX "assignable(void *ptr, "
+    CPROVER_PREFIX "size_t size,"
+    CPROVER_PREFIX "bool is_ptr_to_ptr);\n"
```
Please indent

remi-delmas-3000 replied:
fixed

tautschnig commented on src/ansi-c/ansi_c_internal_additions.cpp:
```diff
@@ -216,6 +216,19 @@ void ansi_c_internal_additions(std::string &code)
+    CPROVER_PREFIX "bool is_ptr_to_ptr);\n"
+    // Declares a range of bytes as assignable
+    "void " CPROVER_PREFIX "object_upto(void *ptr, "
+    CPROVER_PREFIX "size_t size);\n"
```
Please indent

remi-delmas-3000 replied:
fixed

tautschnig commented on src/ansi-c/c_typecheck_base.cpp:
```diff
@@ -768,14 +768,37 @@ void c_typecheck_baset::typecheck_declaration(
-          ID_loop_entry,
-          CPROVER_PREFIX "loop_entry is not allowed in preconditions.");
-      };
+      auto check_history_expr =
+        [&](const exprt expr, std::string &clause_type) {
+          disallow_subexpr_by_id(
+            expr,
+            ID_old,
+            CPROVER_PREFIX "old is not allowed in " + clause_type + ".");
+          disallow_subexpr_by_id(
+            expr,
+            ID_loop_entry,
+            CPROVER_PREFIX "loop_entry is not allowed in " + clause_type + ".");
+        };
+
+      auto check_return_value =
+        [&](const exprt &expr, std::string &clause_type) {
+          const irep_idt id = CPROVER_PREFIX "return_value";
+
+          auto pred = [&](const exprt &expr) {
+            if(!can_cast_expr<symbol_exprt>(expr))
+              return false;
+
+            return to_symbol_expr(expr).get_identifier() == id;
+          };
+
+          if(!has_subexpr(expr, pred))
+            return;
+
+          error().source_location = expr.source_location();
+          error() << id2string(id) + " is not allowed in " << clause_type << "."
+                  << eom;
+          throw 0;
+        };
```
Could these please moved to a (static) function? The body of the encompassing function is way too large already (no fault of your own, of course).

remi-delmas-3000 replied:
fixed

tautschnig commented on src/ansi-c/c_typecheck_base.cpp:
```diff
@@ -822,7 +845,9 @@ void c_typecheck_baset::typecheck_declaration(
           typecheck_spec_function_pointer_obeys_contract(expr);
-          check_history_expr(expr);
+          std::string clause_type = "function pointer preconditions";
+          check_history_expr(expr, clause_type);
+          check_return_value(expr, clause_type);
```
If these always appear in pairs, then they should really be one function to call?!

remi-delmas-3000 replied:
fixed

tautschnig commented on src/ansi-c/c_typecheck_base.cpp:
```diff
@@ -768,14 +768,37 @@ void c_typecheck_baset::typecheck_declaration(
+          error().source_location = expr.source_location();
+          error() << id2string(id) + " is not allowed in " << clause_type << "."
+                  << eom;
+          throw 0;
```
Throw `invalid_source_file_exceptiont`

tautschnig commented on src/ansi-c/c_typecheck_code.cpp:
```diff
@@ -847,7 +847,26 @@ void c_typecheck_baset::typecheck_dowhile(code_dowhilet &code)
+{
+  if(has_subexpr(expr, ID_side_effect))
+  {
+    std::ostringstream error_message;
+    error_message << "side-effects not allowed in assigns clause targets";
+    throw invalid_source_file_exceptiont{
+      error_message.str(), expr.source_location()};
+  }
+  if(has_subexpr(expr, ID_if))
+  {
+    std::ostringstream error_message;
+    error_message << "ternary expressions not allowed in assigns "
+                     "clause targets";
```
Is there really a need to use a stream here?

tautschnig commented on src/ansi-c/c_typecheck_code.cpp:
```diff
@@ -869,80 +888,121 @@ void c_typecheck_baset::typecheck_spec_assigns_condition(exprt &condition)
-    error().source_location = condition.source_location();
-    error() << "void-typed expressions not allowed as assigns clause conditions"
-            << eom;
+    std::ostringstream error_message;
+    error_message
+      << "void-typed expressions not allowed as assigns clause conditions";
+    throw invalid_source_file_exceptiont{
+      error_message.str(), condition.source_location()};
```
Stream not needed?!

remi-delmas-3000 replied:
fixed

tautschnig commented on src/ansi-c/c_typecheck_code.cpp:
```diff
@@ -869,80 +888,121 @@ void c_typecheck_baset::typecheck_spec_assigns_condition(exprt &condition)
-    must_throw = true;
-    error().source_location = condition.source_location();
-    error() << "side-effects not allowed in assigns clause conditions" << eom;
+    std::ostringstream error_message;
+    error_message << "side-effects not allowed in assigns clause conditions";
+    throw invalid_source_file_exceptiont{
+      error_message.str(), condition.source_location()};
```
No need for a stream?!

remi-delmas-3000 replied:
fixed

tautschnig commented on src/ansi-c/c_typecheck_code.cpp:
```diff
@@ -869,80 +888,121 @@ void c_typecheck_baset::typecheck_spec_assigns_condition(exprt &condition)
-    error() << "ternary expressions not allowed in assigns "
-               "clause conditions"
-            << eom;
+    std::ostringstream error_message;
+    error_message
+      << "ternary expressions not allowed in assigns clause conditions";
+    throw invalid_source_file_exceptiont{
+      error_message.str(), condition.source_location()};
```
No need for a stream?!

remi-delmas-3000 replied:
fixed

tautschnig commented on src/ansi-c/c_typecheck_code.cpp:
```diff
@@ -869,80 +888,121 @@ void c_typecheck_baset::typecheck_spec_assigns_condition(exprt &condition)
+      std::ostringstream error_message;
+      error_message << "expected a conditional target group expression in " +
+                         clause_type + "clause, found "
+                    << id2string(target.id());
```
No need for `id2string` on this occasion.

remi-delmas-3000 replied:
fixed

tautschnig commented on src/ansi-c/c_typecheck_code.cpp:
```diff
@@ -869,80 +888,121 @@ void c_typecheck_baset::typecheck_spec_assigns_condition(exprt &condition)
+    if(!can_cast_expr<conditional_target_group_exprt>(target))
+    {
+      std::ostringstream error_message;
+      error_message << "expected a conditional target group expression in " +
+                         clause_type + "clause, found "
```
Use `<<` if you're using a stream, otherwise it defeats the purpose of using a stream.

remi-delmas-3000 replied:
fixed

tautschnig commented on src/ansi-c/c_typecheck_code.cpp:
```diff
@@ -869,80 +888,121 @@ void c_typecheck_baset::typecheck_spec_assigns_condition(exprt &condition)
 {
-  // compute type
-  typecheck_expr(target);
+  exprt::operandst tmp;
```
Although we do it way too many times across the code base: could you try to find a more descriptive name than "tmp?"

remi-delmas-3000 replied:
is `new_targets` ok ?

tautschnig commented on src/ansi-c/c_typecheck_base.h:
```diff
@@ -148,10 +148,15 @@ class c_typecheck_baset:
-  virtual void typecheck_spec_assigns_condition(exprt &condition);
+  virtual void typecheck_conditional_targets(
+    exprt::operandst &targets,
+    const std::function<void(exprt &)> typecheck_target,
```
It seems `typecheck_conditional_targets` is called exactly once, making the need for this configurable function unnecessary?!

remi-delmas-3000 replied:
this is is mean to allow checking both assigns and frees-clauses targets (next PR)

tautschnig commented on src/ansi-c/c_typecheck_code.cpp:
```diff
@@ -869,80 +888,121 @@ void c_typecheck_baset::typecheck_spec_assigns_condition(exprt &condition)
   {
+    if(target.type().id() == ID_empty)
+    {
+      std::ostringstream error_message;
+      error_message
+        << "lvalue expressions with void type not allowed in assigns clauses";
+      throw invalid_source_file_exceptiont{
```
No need for a stream

remi-delmas-3000 replied:
fixed

tautschnig commented on src/ansi-c/c_typecheck_code.cpp:
```diff
@@ -951,38 +1011,33 @@ void c_typecheck_baset::typecheck_spec_assigns_target(exprt &target)
-      error().source_location = target.source_location();
-      error() << "function pointer calls not allowed in assigns targets" << eom;
-      throw 0;
+      std::ostringstream error_message;
+      error_message << "function pointer calls not allowed in assigns clauses";
+      throw invalid_source_file_exceptiont{
+        error_message.str(), target.source_location()};
```
No need for a stream

remi-delmas-3000 replied:
fixed

tautschnig commented on src/ansi-c/c_typecheck_expr.cpp:
```diff
@@ -1947,8 +1947,59 @@ void c_typecheck_baset::typecheck_side_effect_function_call(
       // This is an undeclared function.
+
+      // Is it the polymorphic typed_target function ?
+      if(identifier == CPROVER_PREFIX "typed_target")
+      {
+        if(expr.arguments().size() != 1)
+        {
```
I would appreciate if the body of this `if` statement could be factored out into a function to reduce the overall size (in the number of lines) of the encompassing function.

remi-delmas-3000 replied:
fixed

tautschnig commented on src/ansi-c/c_typecheck_expr.cpp:
```diff
@@ -1947,8 +1947,59 @@ void c_typecheck_baset::typecheck_side_effect_function_call(
+            error_message.str(), expr.source_location()};
+        }
+
+        auto arg0 = expr.arguments().at(0);
```
Nit pick: We don't commonly use `at`, and you've done the sanity check just before, so perhaps use `[0]` or `.front()` instead?

remi-delmas-3000 replied:
fixed

tautschnig commented on src/ansi-c/c_typecheck_expr.cpp:
```diff
@@ -1947,8 +1947,59 @@ void c_typecheck_baset::typecheck_side_effect_function_call(
+        typecheck_expr(arg0);
+        if(!is_assignable(arg0) || !arg0.get_bool(ID_C_lvalue))
+        {
+          std::ostringstream error_message;
+          error_message << "argument of " << CPROVER_PREFIX "typed_target"
+                        << "must be assignable";
+          throw invalid_source_file_exceptiont{
```
You could choose to avoid the stream here as well, the string is a constant.

remi-delmas-3000 replied:
fixed

tautschnig commented on src/goto-instrument/contracts/instrument_spec_assigns.cpp:
```diff
@@ -519,6 +519,44 @@ car_exprt instrument_spec_assignst::create_car_expr(
+      else
+      {
+        log.error().source_location = target.source_location();
+        log.error() << "call to " + id2string(ident) +
+                         " in assigns clauses not supported in "
+                         "this version";
```
Use `<<` instead of `+` (and then you can get rid of `id2string`).

remi-delmas-3000 replied:
fixed

nwetzler commented on src/ansi-c/c_typecheck_code.cpp:
```diff
@@ -847,7 +847,23 @@ void c_typecheck_baset::typecheck_dowhile(code_dowhilet &code)
 }
 
-void c_typecheck_baset::typecheck_spec_assigns_condition(exprt &condition)
+void c_typecheck_baset::throw_on_side_effects(const exprt &expr)
```
Do we have anything in regression testing for side effects in expressions in the context of assigns clauses?

remi-delmas-3000 replied:
Yes side effects in assigns clause targets are rejected and we have tests for that in `regression/contracts/assigns_enforce_side_effects*`

nwetzler commented on doc/cprover-manual/contracts-assigns.md:
```diff
@@ -1,75 +1,373 @@
+by the function (or the loop), even if they end up holding the same value as they
+held before the function call (or before entering the loop).
+
+Memory locations that are locally stack-or heap-allocated during function
```
```suggestion
Memory locations that are locally stack- or heap-allocated during function
```

remi-delmas-3000 replied:
Fixed

nwetzler commented on doc/cprover-manual/contracts-assigns.md:
```diff
@@ -1,75 +1,373 @@
-In order to determine whether an _assigns_ clause is a sound abstraction of
-the write set of a function *f*, the body of the function is instrumented with
+// Proof harness
+int main() {
```
Are these examples all in the regression somewhere?  If so, is there a way we could link to them?

remi-delmas-3000 replied:
No these are not in the regression test suite. I wouldn't know how to link the manual to the regression tests in a persistent way.

nwetzler commented on regression/contracts/reject_history_expr_in_assigns_clause/test.desc:
```diff
@@ -0,0 +1,11 @@
+CORE
+main.c
+--enforce-contract foo
+^main.c.*error: assigns clause target must be a non-void lvalue or a call to one of __CPROVER_POINTER_OBJECT, __CPROVER_assignable, __CPROVER_object_whole, __CPROVER_object_upto, __CPROVER_object_from$
```
Do we still want `__CPROVER_POINTER_OBJECT` here?

remi-delmas-3000 replied:
Yes for compatibility reasons until we merge this one https://github.com/diffblue/cbmc/pull/7127
